apps/admin/backend/src/cvr_files.ts

```diff
@@ -689,7 +606,79 @@ export async function addCastVoteRecordReport({
           index: castVoteRecordIndex,
         });
       }
-      const { id: cvrId, isNew: cvrIsNew } = addCastVoteRecordResult.ok();
+      const { cvrId, isNew: cvrIsNew } = addCastVoteRecordResult.ok();
+
+      if (cvrIsNew) {
+        // Add images to the store
+        const ballotImageIds: Array<Optional<Id>> = [undefined, undefined];
+        if (cvr.BallotImage) {
+          // Convention is that we always have two entries in the BallotImage
+          // array, allowing us to indicate front and back via array index.
+          assert(cvr.BallotImage.length === 2);
+
+          for (const [pageIndex, cvrImageData] of cvr.BallotImage.entries()) {
+            // There may be no image data for the current page
+            if (!cvrImageData.Location) {
+              continue;
+            }
+
+            // Load image data
+            assert(cvrImageData.Location.startsWith('file:'));
+            const imagePath = join(
+              reportDirectoryPath,
+              cvrImageData.Location.slice('file:'.length)
+            );
+            const imageData = await fs.readFile(imagePath);
+
+            // Load and verify layout data
+            const layoutPath = imagePath
+              .replace(
+                CVR_BALLOT_IMAGES_SUBDIRECTORY,
+                CVR_BALLOT_LAYOUTS_SUBDIRECTORY
+              )
+              .replace(/(jpg|jpeg)$/, 'layout.json');
```

Comment:
All of this is a bit fragile and may not do the right thing depending on exactly where on the file system this is. For example, if the USB drive gets mounted at a directory with `ballot-images` in the name this will produce an incorrect path. Similarly, if we ever use non-`jpeg` or `jpg` extensions this will not work correctly. There's an argument to be made that we just append `.layout.json` to whatever filename the image has. However, this function does the same thing yours does but works around these issues and also ensures that we only read files within the `reportDirectoryPath`:

```ts
async function loadImageAndAssociatedLayout(
  cvrImageDataLocation: string,
  reportDirectoryPath: string
): Promise<LoadedCvrImageAndLayout> {
  assert(cvrImageDataLocation.startsWith('file:'));
  const relativeImagePath = normalize(
    cvrImageDataLocation.slice('file:'.length)
  );
  assert(relativeImagePath.startsWith(`${CVR_BALLOT_IMAGES_SUBDIRECTORY}/`));
  const parsedBallotImagePath = parse(
    relativeImagePath.slice(`${CVR_BALLOT_IMAGES_SUBDIRECTORY}/`.length)
  );
  const imagePath = join(reportDirectoryPath, relativeImagePath);
  const imageData = await fs.readFile(imagePath);

  // Load layout data
  const layoutPath = join(
    reportDirectoryPath,
    CVR_BALLOT_LAYOUTS_SUBDIRECTORY,
    parsedBallotImagePath.dir,
    `${parsedBallotImagePath.name}.layout.json`
  );
  const layoutJson = await fs.readFile(layoutPath, 'utf8');

  return { imagePath, layoutPath, imageData, layoutJson };
}
```

Reply:
@eventualbuddha a version of this has been added, which addresses your other security comment.

Reply:
@eventualbuddha I actually think this change is not strictly necessary due to the validation logic upstream of this code. The import code finds all the images _actually_ contained in the report (at the correct locations within the report) and during validation, if the an image reference is not in the list of those _actually_ contained in the report, it will fail. I think this is still a great improvement, however, so I'll keep it in.